R/run-nmtran.R

```diff
@@ -0,0 +1,223 @@
+  # impact the evaluation of `NM-TRAN`
+  .bbi_args <- parse_args_list(.bbi_args, .mod[[YAML_BBI_ARGS]])
+  .bbi_args <- Filter(Negate(is.null), .bbi_args[c("prdefault", "tprdefault", "maxlim")])
+  cmd_args <- check_bbi_args(.bbi_args)
+
+
+  mod_path <- get_model_path(.mod)
+  data_path <- get_data_path_from_ctl(.mod)
+
+  # make temporary directory in current directory
+  mod_name <- fs::path_ext_remove(basename(mod_path))
+  temp_folder <- paste0("nmtran_", mod_name, "_", basename(tempdir()))
+  if(fs::dir_exists(temp_folder)) fs::dir_delete(temp_folder)
+  fs::dir_create(temp_folder)
+  if(isTRUE(delete_on_exit)){
+    on.exit(unlink(temp_folder, recursive = TRUE, force = TRUE))
+  }
+
+  # Copy model
+  file.copy(mod_path, temp_folder, overwrite = TRUE)
+  nmtran_mod <- new_model(file.path(temp_folder, basename(mod_path)), .overwrite = TRUE)
+
+  # Copy dataset & overwrite $DATA record of new model
+  # NMTRAN will error if data cannot be found
+  if(fs::file_exists(data_path)){
+    file.copy(data_path, temp_folder, overwrite = TRUE)
+    # overwrite $DATA record of new model
+    modify_data_path_ctl(nmtran_mod, data_path = basename(data_path))
+  }
+
+  # Run NMTRAN
+  nmtran_results <- c(
+    list(
+      nmtran_exe = as.character(nmtran_exe),
+      nonmem_version = nm_ver,
+      absolute_model_path = mod_path
+    ),
+    execute_nmtran(
+      nmtran_exe, mod_path = basename(mod_path), cmd_args = cmd_args,
+      dir = temp_folder
+    )
+  )
+
+  # assign class and return
+  class(nmtran_results) <- c(NMTRAN_PROCESS_CLASS, class(nmtran_results))
+  return(nmtran_results)
+}
+
+
+#' Search for and validate existence of an `NM-TRAN` executable
+#'
+#' If `nmtran_exe = NULL`, this will look for a `bbi.yaml` file in the same
+#' directory as the model.
+#'
+#' @inheritParams run_nmtran
+#'
+#' @keywords internal
+locate_nmtran <- function(.mod = NULL, .config_path = NULL, nmtran_exe = NULL){
+
+  if(is.null(nmtran_exe)){
+    if(!is.null(.mod)){
+      check_model_object(.mod, "bbi_nonmem_model")
+      model_dir <- get_model_working_directory(.mod)
+    }
+    config_path <- .config_path %||% file.path(model_dir, "bbi.yaml")
+
+    if(!file_exists(config_path)){
+      rlang::abort(
+        c(
+          "x" = "No bbi configuration was found in the execution directory.",
+          "i" = "Please run `bbi_init()` with the appropriate directory to continue."
+        )
+      )
+    }
+
+    if(!is.null(.config_path)){
+      config_path <- normalizePath(.config_path)
+    }
+
+    bbi_config <- yaml::read_yaml(config_path)
+    nm_config <- bbi_config$nonmem
+
+    # look for default nonmem installation
+    default_nm <- purrr::keep(nm_config, function(nm_ver){
+      !is.null(nm_ver$default)
+    })
+
+    # Set nonmem path
+    if(length(default_nm) > 0){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
```

Comment:
"likely higher version" &#x2013; hmm, yeah, this is good enough. It makes a lot of assumptions about how `bbi init` and underlying Go functions work, as well as the naming in `/opt/NONMEM/`.

Another option would be to error if no default is found. I think that's particularly appealing if `run_nmtran` is reworked (as mentioned in another comment) to look for `nm_version` in `.bbi_args`.

Reply:
If you still lean in this direction after the refactor let me know (I opted to go with your suggestion of specifying the `nonmem` version instead of an `NM-TRAN` executable). 

I like the idea of picking an available version rather than aborting since we report the `NONMEM` version used anyways, and am more of the mindset that it's better to "just work" rather than forcing a user to have to go back and update their `bbi` configuration file.

Reply:
I'm good with whatever you choose here.  

> am more of the mindset that it's better to "just work" rather than forcing a user to have to go back and update their `bbi` configuration file.

The flip side of that is: they're going to have to do that anyway when they submit this model.  So aborting has the advantages of 1) signaling the issue earlier and 2) being consistent with submit_mode.